File: DataStructures_CPP/BIT.cpp
template<class Type, int SZ>
class BIT {
public:
    Type fen[SZ * 2];
    int n;

    Type query(int idx) {
        Type ret = 0;
        for(int i = idx + 5; i > 0; i -= i & -i)
            ret += fen[i];
        return ret;
    }

    void update(int idx, Type val) {
        for(int i = idx + 5; i < n * 2; i += i & -i)
            fen[i] += val;
    }
};

File: DataStructures/BIT.java
package DataStructures;

class BIT {
    int[] fen;
    int n;
    public BIT(int n) {
        this.n = n;
        fen = new int[n * 2];
    }
    void update(int num, int v) {
        for(int i = num + n; i < fen.length; i += Integer.lowestOneBit(i)) {
            fen[i]+=v;
        }
    }
    int query(int num) {
        int ret = 0;
        for(int i = num + n; i > 0; i -= Integer.lowestOneBit(i)) {
            ret += fen[i];
        }
        return ret;
    }
}
